File: Python/src/MerchantIntegrationOfflineV3.py
import os
import base64
import hashlib
import json
from Config import Config
from RestClientV3 import RestClientV3

class MerchantIntegrationOfflineV3:
    def __init__(
        self, env, country, partner_id, partner_secret, merchant_id, terminal_id
    ):
        config = Config(
            env,
            country,
            partner_id,
            partner_secret,
            merchant_id,
            terminal_id,
            "",
            "",
            "",
        )
        self.config = config

    # 1. V3 Pos - PosInitateQR : generate QRcode for both MPQR and CPQR
    def PosInitate(self, transactionDetails, paymentMethod,POSDetails):
        try:
            msgID = RestClientV3.randomString(32)
            requestBody = {
                "transactionDetails": transactionDetails,
                "msgID": msgID,
            }
            if paymentMethod != None :
                requestBody["paymentMethod"] = paymentMethod
            if POSDetails != None:
                requestBody["POSDetails"] = POSDetails

            path = self.config.pathPosInitiateV3
            return RestClientV3.post(self.config, path, "application/json",requestBody)
        except OSError as err:
            print("Something else went wrong with message:{0}".format(err))

    # 2. V3 Pos - PosInquire : 
    def PosInquire(self, transactionDetails):
        try:
            msgID = RestClientV3.randomString(32)
            requestBody = {
                "transactionDetails": transactionDetails,
                "msgID": msgID,
            }

            path = self.config.pathPosInquireV3
            if "?" in path:
                path += "&"
            else:
                path += "?"
            params = RestClientV3.http_build_query(requestBody)
            params = params.replace("%5B",".")
            params = params.replace("%5D","")
            path += params
            
            return RestClientV3.get(self.config, path, "application/x-www-form-urlencoded","")
        except OSError as err:
            print("Something else went wrong with message:{0}".format(err))


    # 3. V3 Pos - PosCancel : Cancel transaction
    def PosCancel(self, transactionDetails):
        try:
            msgID = RestClientV3.randomString(32)
            requestBody = {
                "transactionDetails": transactionDetails,
                "msgID": msgID,
        
            }
            path = self.config.pathPosCancelV3
            return RestClientV3.put(self.config, path,"application/json", requestBody)
        except OSError as err:
            print("Something else went wrong with message:{0}".format(err))

    # 4. V3 Pos - PosRefund : Refund transaction
    def PosRefund(self, transactionDetails):
        try:
            msgID = RestClientV3.randomString(32)
            requestBody = {
                "transactionDetails": transactionDetails,
                "msgID": msgID,
            }
            path = self.config.pathPosRefundV3
            return RestClientV3.put(self.config, path,"application/json", requestBody)
        except OSError as err:
            print("Something else went wrong with message:{0}".format(err))


    

File: Net/src/Public/MerchantIntegrationOfflineV3.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using Newtonsoft.Json;

namespace Net.Public
{
    public class MerchantIntegrationOfflineV3
    {

        public MerchantConfiguration MerchantConfiguration { get; }
        public IHttpClient HttpClient { get; }

        public MerchantIntegrationOfflineV3(
            string env = null,
            string country = null,
            string partnerId = null,
            string partnerSecret = null,
            string merchantId = null,
            string terminalId = null,
            IHttpClient httpClient = null
            )
        {
            MerchantConfiguration = new MerchantConfiguration(partnerId, partnerSecret, merchantId, "", "", terminalId, "", env, country);
            HttpClient = httpClient ?? new MerchantHttpClient(MerchantConfiguration);
        }

        public HttpResponseMessage PosInitate(Dictionary<string, dynamic> transactionDetails, Dictionary<string, dynamic> paymentMethod, Dictionary<string, dynamic> POSDetails)
        {

            string msgID = Utils.RandomString(32);

            var requestBody = new
            {
                msgID,
                transactionDetails,
                paymentMethod,
                POSDetails
            };

            var content = CreateStringContent(JsonConvert.SerializeObject(requestBody));
            var uri = MerchantConfiguration.BuildUri(PathName.V3PosPaymentInit);
            var request = new MerchantRequest(uri, HttpMethod.Post, content, pathName: PathName.V3PosPaymentInit);
            return HttpClient.SendRequest(request);

        }


        public HttpResponseMessage PosInquire(Dictionary<string, dynamic> transactionDetails)
        {

            string msgID = Utils.RandomString(32);

            var parametersDictionary = new Dictionary<string, dynamic>();
            parametersDictionary.Add("msgID", msgID);
            parametersDictionary.Add("transactionDetails", transactionDetails);
            var paramStr = Utils.BuildQuery(parametersDictionary);
            var uri = MerchantConfiguration.BuildUri(PathName.V3PosPaymentInquiry, parameters: paramStr);
            var content = CreateStringContent("");
            content.Headers.ContentType = new MediaTypeHeaderValue("application/x-www-form-urlencoded");
            var request = new MerchantRequest(uri, HttpMethod.Get, content, contentType: "application/x-www-form-urlencoded", pathName: PathName.V3PosPaymentInquiry);
            return HttpClient.SendRequest(request);
        }

        public HttpResponseMessage PosCancel(Dictionary<string, dynamic> transactionDetails)
        {
            string msgID = Utils.RandomString(32);
            var requestBody = new
            {
                msgID,
                transactionDetails
            };

            var content = CreateStringContent(JsonConvert.SerializeObject(requestBody));
            var uri = MerchantConfiguration.BuildUri(PathName.V3PosCancel);
            var request = new MerchantRequest(uri, HttpMethod.Put, content, pathName: PathName.V3PosCancel);
            return HttpClient.SendRequest(request);
        }

        public HttpResponseMessage PosRefund(Dictionary<string, dynamic> transactionDetails)
        {
            string msgID = Utils.RandomString(32);
            var requestBody = new
            {
                msgID,
                transactionDetails
            };

            var content = CreateStringContent(JsonConvert.SerializeObject(requestBody));
            var uri = MerchantConfiguration.BuildUri(PathName.V3PosRefund);
            var request = new MerchantRequest(uri, HttpMethod.Put, content, pathName: PathName.V3PosRefund);
            return HttpClient.SendRequest(request);
        }

        /// <summary>
        /// Return StringContent object without CharSet
        /// </summary>
        /// <param name="json">json string</param>
        /// <returns></returns>
        private static StringContent CreateStringContent(string json)
        {
            var content = new StringContent(json, Encoding.UTF8, "application/json");
            content.Headers.ContentType.CharSet = "";
            return content;
        }
    }
}
